File: engine/src/main/java/com/trading/hf/patterns/MVELFunctions.java
package com.trading.hf.patterns;

import com.trading.hf.model.VolumeBar;
import java.util.List;
import java.util.stream.Collectors;

public class MVELFunctions {
    
    // Calculate Standard Deviation over the last 'period' bars for a specific field
    public static double stdev(List<VolumeBar> history, int period, String field) {
        if (history == null || history.size() < 2) return 0.0;
        
        List<Double> values = extractLastN(history, period, field);
        if (values.isEmpty()) return 0.0;

        double mean = values.stream().mapToDouble(Double::doubleValue).average().orElse(0.0);
        double variance = values.stream().mapToDouble(v -> Math.pow(v - mean, 2)).average().orElse(0.0);
        
        return Math.sqrt(variance);
    }

    // Calculate Highest value over the last 'period' bars for a specific field
    public static double highest(List<VolumeBar> history, int period, String field) {
        if (history == null || history.isEmpty()) return 0.0;
        
        List<Double> values = extractLastN(history, period, field);
        return values.stream().mapToDouble(Double::doubleValue).max().orElse(0.0);
    }
    
    // Calculate Max (synonym for highest)
    public static double max(List<VolumeBar> history, int period, String field) {
        return highest(history, period, field);
    }

    // Calculate Lowest value over the last 'period' bars for a specific field
    public static double lowest(List<VolumeBar> history, int period, String field) {
        if (history == null || history.isEmpty()) return 0.0;
        
        List<Double> values = extractLastN(history, period, field);
        return values.stream().mapToDouble(Double::doubleValue).min().orElse(0.0);
    }
    
    // Calculate Min (synonym for lowest)
    public static double min(List<VolumeBar> history, int period, String field) {
        return lowest(history, period, field);
    }
    
    // Calculate Simple Moving Average (SMA)
    public static double sma(List<VolumeBar> history, int period, String field) {
         if (history == null || history.isEmpty()) return 0.0;
         
         List<Double> values = extractLastN(history, period, field);
         return values.stream().mapToDouble(Double::doubleValue).average().orElse(0.0);
    }
    
    // Deprecated alias for compatibility if needed, but 'moving_avg' usually implies SMA
    public static double moving_avg(List<VolumeBar> history, int period, String field) {
        return sma(history, period, field);
    }

    // Custom round with scale
    public static double round(double value, int scale) {
        if (scale >= 0) {
            double factor = Math.pow(10, scale);
            return Math.round(value * factor) / factor;
        } else {
            double factor = Math.pow(10, -scale);
            return Math.round(value / factor) * factor;
        }
    }
    
    // Helper to extract the last N values for a field
    private static List<Double> extractLastN(List<VolumeBar> history, int n, String field) {
        int size = history.size();
        int start = Math.max(0, size - n);
        List<VolumeBar> subList = history.subList(start, size);
        
        return subList.stream().map(bar -> {
            switch (field.toLowerCase()) {
                case "close": return bar.getClose();
                case "high": return bar.getHigh();
                case "low": return bar.getLow();
                case "open": return bar.getOpen();
                case "volume": return (double) bar.getVolume();
                case "atr": return bar.getATR();
                default: return 0.0;
            }
        }).collect(Collectors.toList());
    }
}

File: python_engine/utils/mvel_functions.py
import math
import statistics

def _extract_last_n(history, n, field):
    sub_list = history[-n:]
    return [getattr(bar, field.lower(), 0.0) for bar in sub_list]

def stdev(history, period, field):
    if len(history) < 2:
        return 0.0
    values = _extract_last_n(history, period, field)
    return statistics.stdev(values) if len(values) > 1 else 0.0

def highest(history, period, field):
    if not history:
        return 0.0
    values = _extract_last_n(history, period, field)
    return max(values) if values else 0.0

def lowest(history, period, field):
    if not history:
        return 0.0
    values = _extract_last_n(history, period, field)
    return min(values) if values else 0.0

def moving_avg(history, period, field):
    if not history:
        return 0.0
    values = _extract_last_n(history, period, field)
    return statistics.mean(values) if values else 0.0

def ema(history, period, field):
    if not history:
        return 0.0
    values = _extract_last_n(history, min(len(history), period * 2), field)
    if not values:
        return 0.0
    alpha = 2 / (period + 1)
    ema_val = values[0]
    for val in values[1:]:
        ema_val = val * alpha + ema_val * (1 - alpha)
    return ema_val

def vwap(history):
    if not history:
        return 0.0
    total_pv = 0.0
    total_v = 0.0
    # VWAP is usually reset daily, but for historical context we can calculate over window
    # Assuming history passed is for the current day
    for bar in history:
        tp = (bar.high + bar.low + bar.close) / 3.0
        total_pv += tp * bar.volume
        total_v += bar.volume
    return total_pv / total_v if total_v > 0 else history[-1].close

def rsi(history, period=14):
    if len(history) < period:
        return 50.0
    closes = _extract_last_n(history, period + 1, 'close')
    gains = []
    losses = []
    for i in range(1, len(closes)):
        diff = closes[i] - closes[i-1]
        if diff > 0:
            gains.append(diff)
            losses.append(0)
        else:
            gains.append(0)
            losses.append(abs(diff))
    
    avg_gain = statistics.mean(gains)
    avg_loss = statistics.mean(losses)
    
    if avg_loss == 0:
        return 100.0
    rs = avg_gain / avg_loss
    return 100.0 - (100.0 / (1.0 + rs))

def bb_upper(history, period=20, mult=2.0):
    if len(history) < period:
        return 0.0
    ma = moving_avg(history, period, 'close')
    sd = stdev(history, period, 'close')
    return ma + (mult * sd)

def bb_lower(history, period=20, mult=2.0):
    if len(history) < period:
        return 0.0
    ma = moving_avg(history, period, 'close')
    sd = stdev(history, period, 'close')
    return ma - (mult * sd)

MVEL_FUNCTIONS = {
    "abs": abs,
    "round": round,
    "stdev": stdev,
    "highest": highest,
    "max": highest,
    "lowest": lowest,
    "min": lowest,
    "moving_avg": moving_avg,
    "sma": moving_avg,
    "ema": ema,
    "vwap": vwap,
    "rsi": rsi,
    "bb_upper": bb_upper,
    "bb_lower": bb_lower,
    "high_wick": lambda candle: (candle.high - max(candle.open, candle.close)),
    "low_wick": lambda candle: (min(candle.open, candle.close) - candle.low),
    "body_size": lambda candle: abs(candle.open - candle.close),
    "candle_size": lambda candle: (candle.high - candle.low),
}
